core/app/app.go

```diff
@@ -0,0 +1,213 @@
+
+const (
+	busTimeout        = time.Second * 10
+	addRecordTimeout  = time.Second * 10
+	fetchEventTimeout = time.Second * 15
+)
+
+// App provides a bidirectional hook for thread-based apps.
+type App interface {
+	// LocalEventListen returns a listener that is notified of *locally generated*
+	// events. Caller should call .Discard() when done.
+	LocalEventListen() *LocalEventListener
+
+	// HandleNetRecord handles an inbound thread record from net.
+	HandleNetRecord(rec net.ThreadRecord, key thread.Key, lid peer.ID, timeout time.Duration) error
+}
+
+// LocalEventsBus wraps a broadcaster for local events.
+type LocalEventsBus struct {
+	bus *broadcast.Broadcaster
+}
+
+// NewLocalEventsBus returns a new bus for local event.
+func NewLocalEventsBus() *LocalEventsBus {
+	return &LocalEventsBus{bus: broadcast.NewBroadcaster(0)}
+}
+
+// Send an IPLD node and thread auth into the bus.
+// These are received by the app connector and written to the underlying thread.
+func (leb *LocalEventsBus) Send(node format.Node, auth *thread.Auth) error {
+	return leb.bus.SendWithTimeout(&LocalEvent{Node: node, Auth: auth}, busTimeout)
+}
+
+// Listen returns a local event listener.
+func (leb *LocalEventsBus) Listen() *LocalEventListener {
+	l := &LocalEventListener{
+		listener: leb.bus.Listen(),
+		c:        make(chan *LocalEvent),
+	}
+	go func() {
+		for v := range l.listener.Channel() {
+			events := v.(*LocalEvent)
+			l.c <- events
+		}
+		close(l.c)
+	}()
+	return l
+}
+
+func (leb *LocalEventsBus) Discard() {
+	leb.bus.Discard()
+}
+
+// LocalEvent wraps an IPLD node and auth for delivery to a thread.
+type LocalEvent struct {
+	Node format.Node
+	Auth *thread.Auth
+}
+
+// LocalEventListener notifies about new locally generated events.
+type LocalEventListener struct {
+	listener *broadcast.Listener
+	c        chan *LocalEvent
+}
+
+// Channel returns an unbuffered channel to receive local events.
+func (l *LocalEventListener) Channel() <-chan *LocalEvent {
+	return l.c
+}
+
+// Discard indicates that no further events will be received
+// and ready for being garbage collected.
+func (l *LocalEventListener) Discard() {
+	l.listener.Discard()
+}
+
+// Net adds the ability to connect an app to a thread.
+type Net interface {
+	net.Net
+
+	// ConnectApp returns an app<->thread connector.
+	ConnectApp(App, thread.ID) (*Connector, error)
+}
+
+// Connector connects an app to a thread.
+type Connector struct {
+	app        App
+	net        Net
+	threadID   thread.ID
+	threadKey  thread.Key
+	logID      peer.ID
+	closeChan  chan struct{}
+	goRoutines sync.WaitGroup
+
+	lock   sync.Mutex
+	closed bool
+}
+
+// Connection receives new thread records, which are pumped to the app.
+type Connection func(context.Context, map[thread.ID]struct{}) (<-chan net.ThreadRecord, error)
+
+// NewConnector creates bidirectional connection between an app and a thread.
+func NewConnector(app App, net Net, tinfo thread.Info, conn Connection) (*Connector, error) {
```

Comment:
The app calls `ConnectApp`, which must return a connector. The net layer implements this func, which uses this constructor to create the connection. It simply needs to return the thread that's being connected, and a connection func that is used to notify the app of new records.